util/collections.go

```diff
@@ -172,8 +172,27 @@ func StringListInsert(list []string, element string, index int) []string {
 // KeyValuePairListToMap converts a list of key value pair encoded as `key=value` strings into a map.
 func KeyValuePairStringListToMap(asList []string) (map[string]string, error) {
 	asMap := map[string]string{}
+
 	for _, arg := range asList {
-		parts := strings.Split(arg, "=")
+		var parts []string
+
+		// splits the string to the slice `parts`, using `=` sign as delimiter, but taking into account that the `=` sign can also be used as a git tag, e.g. `git@github.com/test.git?ref=feature`
+		for i, l := 0, 0; i <= len(arg); i++ {
+			if i == len(arg) {
+				parts = append(parts, arg[l:])
+				break
+			}
+
+			if arg[i] == '=' {
+				if i >= 4 && arg[i-4:i] == "?ref" {
```

Comment:
Not sure if in `util/collections.go`, which can be used in multiple places, should be implemented handling of a specific case for "?ref" strings, it may have side effects in other places

Reply:
Can we try to parse map in handling of `optTerragruntSourceMap`?
https://github.com/gruntwork-io/terragrunt/blob/master/cli/args.go#LL101C12-L101C12

Reply:
> Can we try to parse map in handling of `optTerragruntSourceMap`? https://github.com/gruntwork-io/terragrunt/blob/master/cli/args.go#LL101C12-L101C12

In this case, URLs passed via env var `TERRAGRUNT_SOURCE_MAP` will not be parsed `?ref=` as part of the URL and will break the URL at that point. In my opinion, data passed through the flag or the environment variable should be processed the same way. Or have I missed something?
https://github.com/gruntwork-io/terragrunt/blob/1494e777b1f2ce18797ce1c9e46859bdef7f33ed/cli/args.go#L97

> Not sure if in `util/collections.go`, which can be used in multiple places, should be implemented handling of a specific case for "?ref" strings, it may have side effects in other places

I checked, the func is used only in two places and is relevant in both places, the reason described above.
```
cli/args.go:499:14:     return util.KeyValuePairStringListToMap(asList)
cli/args.go:511:14:     return util.KeyValuePairStringListToMap(mappingsAsList)
```

Reply:
For me, looks like the single responsibility of `KeyValuePairStringListToMap()` is broken - instead of splitting string by `=` it is also checking for `?ref` values.

I was wondering if we can prepare input data to be handled correctly by `KeyValuePairStringListToMap()`

Reply:
I don't think so, this function keeps going to do the same task, splitting string to _source_ and _destination_ URL, for example, the func accepts the follow string 
`git@github.com/old.git?ref=some-tag=git@github.com/new.git?ref=another-tag`, and everything it does is to split string at the correct point 
`git@github.com/old.git?ref=some-tag` `git@github.com/new.git?ref=another-tag` instead of 
`git@github.com/old.git?ref` `some-tag` `git@github.com/new.git?ref` `another-tag`. 
So we can't do any preparation, because any preparation requires proper string splitting first.